### PR TITLE
feat(lambda): provisioned concurrency

### LINKED ISSUE #5298: aws-lambda: Provisioned Concurrency support
AWS Lambda announces Provisioned Concurrency
https://aws.amazon.com/tw/about-aws/whats-new/2019/12/aws-lambda-announces-provisioned-concurrency/

CFN resources affected

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-version.html
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-alias.html

### Other

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by nataibi**
fixes #5298 

----

Pull Request Checklist
- [x] **Testing**
Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
CLI change?: coordinate update of integration tests with team
cdk-init template change?: coordinated update of integration tests with team
- [x] **Docs**
jsdocs: All public APIs documented
README: README and/or documentation topic updated
Design: For significant features, design document added to design folder
- [x] **Title and Description**
Change type: title prefixed with fix, feat and module name in parens, which will appear in changelog
Title: use lower-case and doesn't end with a period
Breaking?: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
Issues: Indicate issues fixed via: "Fixes #xxx" or "Closes #xxx"
- [ ] Sensitive Modules (requires 2 PR approvers)
IAM Policy Document (in @aws-cdk/aws-iam)
EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
Grant APIs (only if not based on official documentation with a reference)
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by nataibi** [comment]
requires the latest Clouformation resource spec , ResourceSpecificationVersion: 10.0.0

**Comment by nataibi** [comment]
@nija-at thanks for reviewing this, i have made the necessary changes as requested

File: packages/@aws-cdk/aws-lambda/lib/alias.ts [hunk 1]
```diff
@@ -59,6 +59,13 @@ export interface AliasProps {
    * @default No additional versions
    */
   readonly additionalVersions?: VersionWeight[];
+
+  /**
+   * Specifies a provisioned concurrency configuration for a function's version.
+   *
+   * @default No execution number
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda/lib/alias.ts]: Change to "No provisioned concurrency"

File: packages/@aws-cdk/aws-lambda/lib/alias.ts [hunk 2]
```diff
@@ -59,6 +59,13 @@ export interface AliasProps {
    * @default No additional versions
    */
   readonly additionalVersions?: VersionWeight[];
+
+  /**
+   * Specifies a provisioned concurrency configuration for a function's version.
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda/lib/alias.ts]: "function's **alias**"

File: packages/@aws-cdk/aws-lambda/lib/function.ts [hunk 3]
```diff
@@ -541,13 +541,15 @@ export class Function extends FunctionBase {
    * @param codeSha256 The SHA-256 hash of the most recently deployed Lambda source code, or
    *  omit to skip validation.
    * @param description A description for this version.
+   * @param provisionedConcurrencyConfiguration A provisioned concurrency configuration for a function's version.
    * @returns A new Version object.
    */
-  public addVersion(name: string, codeSha256?: string, description?: string): Version {
+  public addVersion(name: string, codeSha256?: string, description?: string, provisionedExecutions?: number): Version {
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda/lib/function.ts]: Keep the doc and code consistent.

File: packages/@aws-cdk/aws-lambda/lib/lambda-version.ts [hunk 4]
```diff
@@ -41,6 +41,13 @@ export interface VersionProps {
    * Function to get the value of
    */
   readonly lambda: IFunction;
+
+  /**
+   * Specifies a provisioned concurrency configuration for a function's version.
+   *
+   * @default ProvisionedConcurrencyConfig execution number
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda/lib/lambda-version.ts]: What's the default?

File: packages/@aws-cdk/aws-lambda/lib/alias.ts [hunk 5]
```diff
@@ -200,6 +208,23 @@ export class Alias extends QualifiedFunctionBase implements IAlias {
       throw new Error(`Sum of additional version weights must not exceed 1, got: ${total}`);
     }
   }
+
+  /**
+   * Validate that the provisionedConcurrentExecutions makes sense
+   *
+   * Member must have value greater than or equal to 1
+   */
+  private determineProvisionedConcurrentExecutions(props: AliasProps) {
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda/lib/alias.ts]: specify a return type

File: packages/@aws-cdk/aws-lambda/lib/lambda-version.ts [hunk 6]
```diff
@@ -155,6 +163,23 @@ export class Version extends QualifiedFunctionBase implements IVersion {
       ...props
     });
   }
+
+  /**
+   * Validate that the provisionedConcurrentExecutions makes sense
+   *
+   * Member must have value greater than or equal to 1
+   */
+  private determineProvisionedConcurrentExecutions(props: VersionProps) {
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda/lib/lambda-version.ts]: specify the return type.

**Comment by nataibi** [comment]
Thanks for clarifying,  let me know if there is further refactoring on this @nija-at , the new changes have been pushed

**Comment by nija-at** [comment]
> We should allow the user to specify provisionedConcurrency on the Function class which wires it up with the $LATEST.

Could you incorporate this comment?

**Comment by nataibi** [comment]
added integration tests, @nija-at thanks for the guidance

**Comment by nija-at** [comment]
I've pushed a change to fix up a couple of indentation issues.

**Pull request merged**